xbmc/pvr/addons/PVRClients.cpp

```diff
@@ -1520,3 +1531,80 @@ bool CPVRClients::IsRealTimeStream(void) const
   return false;
 }
 
+void CPVRClients::ConnectionStateChange(int clientId, std::string &strConnectionString, PVR_CONNECTION_STATE newState,
+                                        std::string &strMessage)
+{
+  PVR_CLIENT client;
+  if (!GetClient(clientId, client))
+  {
+    CLog::Log(LOGDEBUG, "PVR - %s - invalid client id", __FUNCTION__);
+    return;
+  }
+
+  if (strConnectionString.empty())
+  {
+    CLog::Log(LOGERROR, "PVR - %s - invalid handler data", __FUNCTION__);
+    return;
+  }
+
+  int iMsg(-1);
+  bool bError(true);
+  bool bNotify(true);
+
+  switch (newState)
+  {
+    case PVR_CONNECTION_STATE_SERVER_UNREACHABLE:
+      iMsg = 35505; // Server is unreachable
+      break;
+    case PVR_CONNECTION_STATE_SERVER_MISMATCH:
+      iMsg = 35506; // Server does not respond properly
+      break;
+    case PVR_CONNECTION_STATE_VERSION_MISMATCH:
+      iMsg = 35507; // Server version is not compatible
+      break;
+    case PVR_CONNECTION_STATE_ACCESS_DENIED:
+      iMsg = 35508; // Access denied
+      break;
+    case PVR_CONNECTION_STATE_CONNECTED:
+      bError = false;
+      iMsg = 36034; // Connection established
+      break;
```

Comment:
> In any state other than "connecting" PVRManager will try to manage the addon

Wouldn't it be better if pvr manager would only try to manage add-ons in state "connected"? Every other state represents a non-working add-on and it makes no sense to manage it in this state, right?

For backward compatibility it could be necessary to also allow managing add-ons in state "unknown", for the add-ons not supporting async connect. As long as add-ons supporting async connect set "connecting" in AddonDll::Create() all should work just fine, imo.